rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,236 @@
+
+For each fulltext search index a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field in the index.
+
+The proposed process of adding a fulltext search index to an Entity 
+involves adding a special field to that Entity in the GraphQL schema of 
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@index` directive. 
+The directive will have two required parameters for specifying the index
+`type` and `params`. Params fields are specific to the index type; in the 
+case of fulltext search there will be two params to specify: `language` 
+and `fields`. Verification of the index definition will ensure that all 
+fields referenced are valid String type fields. To be clear, the proposed 
+interface will add the fulltext search index definitions to the subgraph 
+deployment hash, so adding or updating a fulltext search index will require 
+an update to the subgraph.  With subgraph composition it will be possible 
+to easily create new subgraphs that add specific fulltext search indexes 
+to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @index(
+        type: "fulltext", 
+        params: {
+            language:"english", 
+            fields: [
+                {name: "name", weight: 5}, 
+                {name: "description", weight: 3}, 
+                {name: "specs", weight: 1},                
+            ],            
+        }
+    )
+}
+```  
+
+### GraphQL Query interface
+
+For each fulltext index three search algorithms will be made available to 
+the end user: a simple text query, a standard ranking, and a cover 
+density ranking algorithm: 
+  - `query`: search for the string in the indexed document. Several 
+    operators are available: and, or, and proximity (`&`, `|`, `<->`.) 
+    The proximity operator allows one to specify max distance: 
+    3 words apart → `<3>`. (no filter suffix)
+  - `standard ranking`: ranking based on the number of matching lexemes. 
+    (filter suffix = `_rank`)
+  - `cover density ranking`: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.
+    (filter suffix = `_proximity_rank`)
+    
+Example Usage in a query (references the example Entity definition from above section): 
+```graphql
+query {
+    exampleItems(first: 10, where: {itemSearch_proximity_rank: "sarcastic&tea"}) {
+        address
+        name 
+        description
+    }
+}
```

Comment:
With my above suggestion, the schema-wide index name could be used as the query field. A fulltext search with the name `search` could be used as follows in queries:
```graphql
{
  ranked: search(first: 10, by: { rank: "sarcastic&tea" }, orderBy: rank, orderDirection: desc) {
    rank
    entity {
      ... on ExampleItem {
        address
        name
      }
    }
  }

  text: search(first: 10, by: { text: "foobar" }) {
    ...
    entity {
      ... on ExampleItem {
        address
        name
      }
    }
  }
}
```

One question I have is how much information the return values should carry in addition to the entity data.